File: frontend/src/services/foods.ts
import { apiGet } from './api'

// --- Backend response types ---

export type FoodItemRaw = {
  id: string
  food_name: string
  weight: string
  calories: number
  protein: number
  carbohydrate: number
  fat: number
  category?: string | null
  confidence?: number | null
  source?: string | null
  estimated?: boolean | null
  image_url?: string | null
  dish_family?: string | null
  alternatives?: Array<{ name: string; confidence?: number | null }> | null
  user_correction?: string | null
  components?: Array<{
    name: string
    confidence?: number | null
    estimated_weight_g?: number | null
    calories?: number | null
    protein?: number | null
    carbs?: number | null
    fat?: number | null
    role?: string | null
    include_in_total?: boolean | null
  }> | null
}

export type FoodRecordRaw = {
  id: string
  status?: string | null
  status_label: string
  confirmed_at?: string | null
  total_calories: number
  protein: number
  carbohydrate: number
  fat: number
  target_calories: number
  image_url?: string | null
  created_at?: string | null
  summary?: string | null
  ocr_text?: string | null
  is_food_detected?: boolean | null
  non_food_reason?: string | null
  analysis_mode?: string | null
}

export type AiLogRaw = {
  prompt_version?: string | null
  latency?: string | null
  cache_hit: boolean
}

export type MacroTargetsRaw = {
  protein: number
  carbs: number
  fat: number
}

export type MacroPercentagesRaw = {
  protein: number
  carbs: number
  fat: number
}

export type FoodDetailRaw = {
  record: FoodRecordRaw
  food_items: FoodItemRaw[]
  ai_log: AiLogRaw
  macro_targets: MacroTargetsRaw
  macro_percentages: MacroPercentagesRaw
}

export type FoodDetailResponseRaw = {
  data: FoodDetailRaw
}

// --- Frontend page types ---

export type MacroInfo = {
  key: 'protein' | 'carbs' | 'fat'
  label: string
  value: number
  target: number
  unit: string
  percent: number
}

export type FoodItem = {
  id: string
  name: string
  weight: string
  calories: number
  protein: number
  carbs: number
  fat: number
  category: string
  confidence: number
  source: string
  estimated: boolean
  imageUrl: string
  includeInTotal?: boolean
  dishFamily?: string | null
  alternatives?: Array<{ name: string; confidence?: number | null }> | null
  userCorrection?: string | null
  components?: Array<{
    name: string
    confidence?: number | null
    estimatedWeightG?: number | null
    calories?: number | null
    protein?: number | null
    carbs?: number | null
    fat?: number | null
    role?: string | null
    includeInTotal?: boolean | null
  }> | null
}

export type AnalyzeResult = {
  id: string
  status: string
  statusLabel: string
  analysisMode: string
  totalCalories: number
  remainingCalories: number
  targetCalories: number
  imageUrl: string
  createdAt: string
  macros: MacroInfo[]
  foodItems: FoodItem[]
  aiSummary: string[]
  technical: {
    ocrText: string
    promptVersion: string
    aiLatency: string
    cacheHit: boolean
  }
}

// --- Adapter ---

function clampPercent(value: number): number {
  return Math.max(0, Math.min(100, Math.round(value)))
}

function resolveImageUrl(value?: string | null): string {
  if (!value) return ''
  const API_BASE = process.env.NEXT_PUBLIC_API_BASE_URL ?? ''
  if (value.startsWith('http://') || value.startsWith('https://')) return value
  if (value.startsWith('/') && API_BASE) return `${API_BASE}${value}`
  return value
}

function splitSummary(raw?: string | null): string[] {
  if (!raw) return []
  return raw
    .split(/\n/)
    .map((s) => s.trim())
    .filter(Boolean)
}

export function adaptFoodDetail(data: FoodDetailRaw): AnalyzeResult {
  const r = data.record
  const macros: MacroInfo[] = [
    {
      key: 'protein', label: '蛋白质', value: r.protein,
      target: data.macro_targets.protein ?? 120, unit: 'g',
      percent: clampPercent(data.macro_percentages?.protein ?? 0),
    },
    {
      key: 'carbs', label: '碳水', value: r.carbohydrate,
      target: data.macro_targets.carbs ?? 250, unit: 'g',
      percent: clampPercent(data.macro_percentages?.carbs ?? 0),
    },
    {
      key: 'fat', label: '脂肪', value: r.fat,
      target: data.macro_targets.fat ?? 70, unit: 'g',
      percent: clampPercent(data.macro_percentages?.fat ?? 0),
    },
  ]

  const totalCalories = r.total_calories ?? 0
  const targetCalories = r.target_calories ?? 2000

  return {
    id: r.id,
    status: r.status ?? 'draft',
    statusLabel: r.status_label ?? '分析完成',
    analysisMode: r.analysis_mode ?? 'dish_with_components',
    totalCalories,
    remainingCalories: Math.max(targetCalories - totalCalories, 0),
    targetCalories,
    imageUrl: resolveImageUrl(r.image_url),
    createdAt: r.created_at ?? '',
    macros,
    foodItems: data.food_items.map((item) => ({
      id: item.id,
      name: item.food_name,
      weight: item.weight,
      calories: item.calories,
      protein: item.protein,
      carbs: item.carbohydrate,
      fat: item.fat,
      category: item.category ?? 'unknown',
      confidence: item.confidence ?? 0,
      source: item.source || 'unknown',
      estimated: item.estimated ?? false,
      imageUrl: resolveImageUrl(item.image_url),
      dishFamily: item.dish_family ?? null,
      alternatives: item.alternatives?.map((a: Record<string, unknown>) => ({ name: String(a.name || ""), confidence: Number(a.confidence) || null })) ?? null,
      userCorrection: item.user_correction ?? null,
      components: item.components?.map(c => ({
        name: c.name, confidence: c.confidence, estimatedWeightG: c.estimated_weight_g,
        calories: c.calories, protein: c.protein, carbs: c.carbs, fat: c.fat,
        role: c.role, includeInTotal: c.include_in_total,
      })) ?? null,
    })),
    aiSummary: splitSummary(r.summary),
    technical: {
      ocrText: r.ocr_text ?? '',
      promptVersion: data.ai_log.prompt_version ?? '',
      aiLatency: data.ai_log.latency ?? '',
      cacheHit: data.ai_log.cache_hit ?? false,
    },
  }
}

export async function getFoodRecord(id: string): Promise<AnalyzeResult> {
  const res = await apiGet<FoodDetailResponseRaw>(`/api/foods/${id}`)
  return adaptFoodDetail(res.data)
}

File: backend/app/schemas/foods.py
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, field_validator

MEAL_TYPE_VALUES = Literal["breakfast", "lunch", "dinner", "snack"]
VALID_MEAL_TYPES = {"breakfast", "lunch", "dinner", "snack"}
VALID_IMAGE_TYPES = {"image/jpeg", "image/png", "image/jpg", "image/webp"}


class FoodUploadResponse(BaseModel):
    task_id: str


class FoodComponentDetail(BaseModel):
    name: str
    confidence: float = 0.5
    estimated_weight_g: Optional[float] = None
    calories: Optional[float] = None
    protein: Optional[float] = None
    carbs: Optional[float] = None
    fat: Optional[float] = None
    role: str = "ingredient"
    include_in_total: bool = False


class FoodItemDetail(BaseModel):
    id: str
    food_name: str
    weight: str
    calories: int
    protein: int
    carbohydrate: int  # DB 字段 carbs，API 返回 carbohydrate
    fat: int
    image_url: Optional[str] = None
    category: str = "unknown"
    confidence: float = 0.0
    source: str = "unknown"
    estimated: bool = False
    components: list[FoodComponentDetail] = []


class FoodRecordDetail(BaseModel):
    id: str
    status: str = "draft"
    status_label: str
    confirmed_at: Optional[datetime] = None
    total_calories: int
    protein: int
    carbohydrate: int
    fat: int
    target_calories: int
    image_url: Optional[str] = None
    created_at: Optional[datetime] = None
    summary: Optional[str] = None
    ocr_text: Optional[str] = None
    is_food_detected: Optional[bool] = None
    non_food_reason: Optional[str] = None


class AiLogDetail(BaseModel):
    prompt_version: Optional[str] = None
    latency: Optional[str] = None
    cache_hit: bool = False


class MacroTargets(BaseModel):
    protein: int
    carbs: int
    fat: int


class MacroPercentages(BaseModel):
    protein: int
    carbs: int
    fat: int


class FoodDetailData(BaseModel):
    record: FoodRecordDetail
    food_items: list[FoodItemDetail]
    ai_log: AiLogDetail
    macro_targets: MacroTargets
    macro_percentages: MacroPercentages


class FoodDetailResponse(BaseModel):
    data: FoodDetailData


class FoodListItemDetail(BaseModel):
    food_name: str


class FoodListItem(BaseModel):
    id: str
    status: str = "draft"
    meal_type: str
    title: str
    time: str
    created_at: Optional[str] = None
    total_calories: int
    summary: Optional[str] = None
    protein: int
    carbohydrate: int
    fat: int
    image_url: Optional[str] = None
    foods: str = ""
    food_items: list[FoodListItemDetail] = []


class FoodListResponse(BaseModel):
    data: list[FoodListItem]


class ConfirmFoodItem(BaseModel):
    id: str
    food_name: str
    weight: str = ""
    category: str = "unknown"
    calories: int = 0
    protein: int = 0
    carbs: int = 0
    fat: int = 0


class DraftFoodItemUpdate(BaseModel):
    """draft update 接受的 item 字段 — 比 ConfirmFoodItem 更宽松"""
    id: str | None = None
    food_name: str | None = None
    display_name: str | None = None
    name: str | None = None
    weight: str | None = None
    estimated_weight_g: float | None = None
    quantity_description: str | None = None
    category: str | None = None
    is_new: bool = False
    name_changed: bool = False


class ConfirmFoodRequest(BaseModel):
    items: list[ConfirmFoodItem]


class ConfirmFoodResponse(BaseModel):
    message: str = "确认成功"
    record_id: str